blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp

```diff
@@ -0,0 +1,363 @@
+            const std::size_t nSamplesToCopy          = std::min(minSamplesBeforeSyncTag, minSamplesOut);
+            if (_isStreamSynchronized && nSamplesToCopy > 0UZ) { // all streams are in sync -> write sample before first Sync tag
+                for (std::size_t i = 0; i < nPorts; i++) {
+                    std::ranges::copy_n(ins[i].begin(), static_cast<std::ptrdiff_t>(nSamplesToCopy), outs[i].begin());
+
+                    publishDroppedSamplesTagIfNotZero(outs[i], _nDroppedSamples[i]);
+                    publishInputTags(ins[i], outs[i], 0UZ, nSamplesToCopy);
+
+                    _nDroppedSamples[i] = 0UZ;
+
+                    ins[i].consume(nSamplesToCopy);
+                    ins[i].consumeTags(nSamplesToCopy);
+                    outs[i].publish(nSamplesToCopy);
+                }
+            } else { // streams are NOT in sync -> check back pressure and drop samples if needed
+                for (std::size_t i = 0; i < nPorts; i++) {
+                    const std::size_t nSamplesToDrop = ins[i].size() < static_cast<std::size_t>(max_history_size) ? 0UZ : ins[i].size() - static_cast<std::size_t>(max_history_size);
+                    if (nSamplesToDrop != 0UZ) {
+                        ins[i].consume(nSamplesToDrop);
+                        ins[i].consumeTags(nSamplesToDrop);
+
+                        _nDroppedSamples[i] += nSamplesToDrop;
+                        outs[i].publish(0UZ);
+                        _isStreamSynchronized = false;
+                    }
+                }
+            }
+        }
+        for (std::size_t i = 0; i < nPorts; i++) {
+            if (!ins[i].isConsumeRequested()) {
+                ins[i].consume(0UZ);
+                ins[i].consumeTags(0UZ);
+                outs[i].publish(0UZ);
+            }
+        }
+        return gr::work::Status::OK;
+    }
+
+private:
+    constexpr void publishDroppedSamplesTagIfNotZero(OutputSpanLike auto& out, std::size_t nDroppedSamples) {
+        if (nDroppedSamples > 0UZ) {
+            out.publishTag({{gr::tag::N_DROPPED_SAMPLES.shortKey(), nDroppedSamples}}, 0UZ);
+        }
+    }
+
+    constexpr void publishInputTags(InputSpanLike auto& in, OutputSpanLike auto& out, std::size_t nDroppedSamples, std::size_t nSamplesToPublish) {
+        const auto tags = in.tags();
+        for (const auto& tag : tags) {
+            if (tag.first >= nDroppedSamples && tag.first < nSamplesToPublish + nDroppedSamples) {
+                out.publishTag(tag.second, tag.first - nDroppedSamples);
+            }
+        }
+    }
+
+    template<InputSpanLike TInput>
+    [[nodiscard]] constexpr std::vector<SyncData> synchronize(const std::span<TInput>& ins) {
+        const std::uint64_t syncTime = findSyncTime(ins);
+        if (syncTime == std::numeric_limits<std::uint64_t>::max()) {
+            return {};
+        }
+
+        std::vector<SyncData> syncData;
```

Comment:
At the risk that this might be a PMO: maybe we should consider making
```cpp
std::vector<SyncData> _syncData;
```
since this is called for every invocation of  `processBulk(..)? similarly for the `findSyncTime(...)` routine below. 

This may be likely a PMO but I am a bit concerned about two vectors being allocated everytime the block is invoked even if there are no synchronisation necessary. Maybe a basic benchmark (w/ and one w/o this block but a simple copy) that measures how much performance is lost might be useful. This way we could eliminate also future doubts and potential regressions.

In terms of benchmarks... guess we have plenty of `bm_,,,` to choose and copy-paste from. :-)